internal/model/transform.go

```diff
@@ -326,6 +339,113 @@ func (m *Model) instantiateStruct(
 	return newType, nil
 }
 
+func (m *Model) instantiateChoice(
+	pkgScope *ast.Package,
+	templChoice *ast.Choice,
+	templateArguments []*ast.TypeReference,
+	instantiatedName string) (*ast.TypeReference, error) {
+	if len(templateArguments) == 0 {
+		return nil, errors.New("choice instantiation called without parameters")
+	}
+	if len(templChoice.TemplateParameters) != len(templateArguments) {
+		return nil, ErrIncorrectNumberOfTemplateParameters
+	}
+	// templateTypes is a map from the template parameter name to the instantiated
+	// type.
+	templateTypes := map[string]*ast.TypeReference{}
+	for ix, ta := range templateArguments {
+		// Instantiate the subtype
+		ta, err := m.instantiateType(pkgScope, ta, ta.TemplateArguments, "")
+		if err != nil {
+			return nil, err
+		}
+		templateTypes[templChoice.TemplateParameters[ix]] = ta
+	}
+
+	newType := &ast.TypeReference{
+		IsBuiltin: false,
+		Package:   pkgScope.Name,
+		Name:      instantiatedName,
+	}
+	if _, exists := pkgScope.Choices[newType.Name]; !exists {
+		choice := &ast.Choice{
+			Name:           newType.Name,
+			Comment:        templChoice.Comment,
+			Cases:          []*ast.ChoiceCase{},
+			Expression:     templChoice.Expression,
+			TypeParameters: []*ast.Parameter{},
+		}
+
+		// resolve the templates for the choices
+		for _, choiceCase := range templChoice.Cases {
+			newChoiceCase := &ast.ChoiceCase{
+				Conditions: choiceCase.Conditions,
+				Comment:    choiceCase.Comment,
+			}
+			if choiceCase.Field != nil {
+				newField, err := m.instantiateField(pkgScope, choiceCase.Field, templateTypes)
+				if err != nil {
+					return nil, err
+				}
+				newChoiceCase.Field = newField
+			}
+			choice.Cases = append(choice.Cases, newChoiceCase)
+
+		}
+
+		// also resolve the default case
+		if templChoice.DefaultCase != nil {
+			choice.DefaultCase = &ast.ChoiceCase{
+				Conditions: templChoice.DefaultCase.Conditions,
+				Comment:    templChoice.DefaultCase.Comment,
+			}
+			if templChoice.DefaultCase.Field != nil {
+				newField, err := m.instantiateField(pkgScope, templChoice.DefaultCase.Field, templateTypes)
+				if err != nil {
+					return nil, err
+				}
+				choice.DefaultCase.Field = newField
+			}
+
+		}
+		// resolve the templates for the types
+		for _, typeParam := range templChoice.TypeParameters {
+			if tt := templateTypes[typeParam.Type.Name]; tt != nil && typeParam.Type.Package == "" {
```

Comment:
similar to my comment above about handling of maps.

Reply:
done